File: dub.js
#!/usr/bin/env node

/**
 * Sarvam AI Video Dubbing CLI Tool (Node.js)
 * No authentication required!
 *
 * Usage:
 *   node dub.js video.mp4 -t Hindi
 *   node dub.js video.mp4 -t Hindi,Tamil --genre podcast
 */

const fs = require("fs");
const path = require("path");
const https = require("https");
const http = require("http");

const BASE_URL = "https://dashboard.sarvam.ai";
const API_DUBBING = `${BASE_URL}/api/dubbing`;
const POLL_INTERVAL = 5000;

const LANGUAGES = [
  "Assamese", "Bengali", "English", "Gujarati", "Hindi",
  "Kannada", "Malayalam", "Marathi", "Odia", "Punjabi",
  "Tamil", "Telugu",
];

const GENRES = [
  "podcast", "monologue", "advertisement",
  "ott_movie_sequence", "edtech", "academic_lecture_formal",
];

// --- HTTP helpers ---

function request(url, options = {}) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const mod = parsed.protocol === "https:" ? https : http;
    const req = mod.request(parsed, {
      method: options.method || "GET",
      headers: options.headers || {},
    }, (res) => {
      const chunks = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => {
        const body = Buffer.concat(chunks).toString();
        resolve({ status: res.statusCode, body, headers: res.headers });
      });
    });
    req.on("error", reject);
    if (options.body) req.write(options.body);
    req.end();
  });
}

function uploadFile(url, filePath) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const mod = parsed.protocol === "https:" ? https : http;
    const fileSize = fs.statSync(filePath).size;
    const stream = fs.createReadStream(filePath);

    const req = mod.request(parsed, {
      method: "PUT",
      headers: {
        "x-ms-blob-type": "BlockBlob",
        "Content-Length": fileSize,
      },
    }, (res) => {
      const chunks = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => resolve({ status: res.statusCode }));
    });
    req.on("error", reject);
    stream.pipe(req);
  });
}

function downloadFile(url, outputPath) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const mod = parsed.protocol === "https:" ? https : http;

    const doDownload = (downloadUrl) => {
      const p = new URL(downloadUrl);
      const m = p.protocol === "https:" ? https : http;
      m.get(p, (res) => {
        if (res.statusCode === 301 || res.statusCode === 302) {
          return doDownload(res.headers.location);
        }
        const total = parseInt(res.headers["content-length"] || "0", 10);
        let downloaded = 0;
        const file = fs.createWriteStream(outputPath);
        res.on("data", (chunk) => {
          downloaded += chunk.length;
          file.write(chunk);
          if (total) {
            process.stdout.write(`\r  ${((downloaded / total) * 100).toFixed(1)}%`);
          }
        });
        res.on("end", () => {
          file.end();
          console.log(`\n  Saved: ${outputPath}`);
          resolve(outputPath);
        });
        res.on("error", reject);
      }).on("error", reject);
    };

    doDownload(url);
  });
}

// --- API functions ---

async function createJob(srcLang, targetLangs, jobName, numSpeakers, genre) {
  const payload = {
    src_lang: srcLang,
    target_langs: targetLangs,
    job_name: jobName,
    num_speakers: numSpeakers,
    genre: genre,
    editor_flow: false,
  };

  console.log(`\nCreating dubbing job...`);
  console.log(`  Source:   ${srcLang}`);
  console.log(`  Target:   ${targetLangs.join(", ")}`);
  console.log(`  Speakers: ${numSpeakers}`);
  console.log(`  Genre:    ${genre}`);

  const resp = await request(`${API_DUBBING}/jobs`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (resp.status !== 200) {
    console.error(`Error creating job: ${resp.status}`);
    console.error(resp.body);
    process.exit(1);
  }

  const data = JSON.parse(resp.body).data;
  console.log(`  Job ID:   ${data.job_id}`);
  return data;
}

async function upload(uploadUrl, videoPath) {
  const size = fs.statSync(videoPath).size;
  console.log(`\nUploading: ${videoPath} (${(size / 1024 / 1024).toFixed(1)} MB)`);

  const resp = await uploadFile(uploadUrl, videoPath);
  if (resp.status !== 200 && resp.status !== 201) {
    console.error(`Error uploading: ${resp.status}`);
    process.exit(1);
  }
  console.log("  Upload complete!");
}

async function startJob(jobId) {
  console.log(`\nStarting job...`);
  const resp = await request(`${API_DUBBING}/jobs/${jobId}/start`, { method: "POST" });
  if (resp.status !== 200) {
    console.error(`Error starting job: ${resp.status}`);
    console.error(resp.body);
    process.exit(1);
  }
  console.log("  Processing started!");
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function pollStatus(jobId) {
  console.log(`\nWaiting for dubbing to complete...`);
  let lastStep = "";

  while (true) {
    const resp = await request(`${API_DUBBING}/jobs/${jobId}/live-status`);
    if (resp.status !== 200) {
      console.log(`  Status check failed: ${resp.status}`);
      await sleep(POLL_INTERVAL);
      continue;
    }

    const data = JSON.parse(resp.body).data || {};
    const status = data.status || "unknown";
    const progress = data.progress || 0;
    const stepLabel = data.current_step_label || "";

    if (stepLabel && stepLabel !== lastStep) {
      console.log(`  [${String(Math.round(progress)).padStart(3)}%] ${stepLabel}`);
      lastStep = stepLabel;
    }

    if (status === "completed") {
      const exp = data.export || {};
      if (exp.status === "completed") {
        console.log(`\n  Dubbing completed!`);
        return data;
      }
      if (exp.status === "failed") {
        console.error(`\n  Export failed: ${data.error_message || "Unknown"}`);
        process.exit(1);
      }
    }

    if (status === "failed") {
      console.error(`\n  Failed: ${data.error_message || "Unknown"}`);
      process.exit(1);
    }

    if (status === "partial_failure") {
      console.error(`\n  Partial failure: ${data.error_message || "Unknown"}`);
      process.exit(1);
    }

    await sleep(POLL_INTERVAL);
  }
}

// --- Arg parser ---

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = {
    video: null,
    srcLang: "English",
    targetLang: null,
    speakers: 1,
    genre: "podcast",
    jobName: null,
    output: null,
    noDownload: false,
    status: null,
    list: false,
  };

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    switch (arg) {
      case "-t": case "--target-lang":
        opts.targetLang = args[++i]; break;
      case "-s": case "--src-lang":
        opts.srcLang = args[++i]; break;
      case "--speakers":
        opts.speakers = parseInt(args[++i], 10); break;
      case "-g": case "--genre":
        opts.genre = args[++i]; break;
      case "--job-name":
        opts.jobName = args[++i]; break;
      case "-o": case "--output":
        opts.output = args[++i]; break;
      case "--no-download":
        opts.noDownload = true; break;
      case "--status":
        opts.status = args[++i]; break;
      case "-l": case "--list":
        opts.list = true; break;
      case "-h": case "--help":
        printHelp(); process.exit(0);
      default:
        if (!arg.startsWith("-")) opts.video = arg;
    }
  }
  return opts;
}

function printHelp() {
  console.log(`
Sarvam AI Video Dubbing CLI (Node.js)

Usage: node dub.js <video> -t <language> [options]

Arguments:
  video                    Path to video file

Options:
  -t, --target-lang LANG   Target language(s), comma-separated (required)
  -s, --src-lang LANG      Source language (default: English)
  --speakers N             Number of speakers (default: 1)
  -g, --genre GENRE        Video genre (default: podcast)
  --job-name NAME          Custom job name
  -o, --output PATH        Output file path
  --no-download            Just print the URL
  --status JOB_ID          Check status of existing job
  -l, --list               List languages & genres
  -h, --help               Show this help

Languages: ${LANGUAGES.join(", ")}
Genres:    ${GENRES.join(", ")}

Examples:
  node dub.js video.mp4 -t Hindi
  node dub.js video.mp4 -t Hindi,Tamil,Telugu --genre edtech
  node dub.js video.mp4 -t Bengali --speakers 2
  node dub.js --status JOB_ID
  `);
}

// --- Main ---

async function main() {
  const opts = parseArgs();

  if (opts.list) {
    console.log(`\nLanguages: ${LANGUAGES.join(", ")}`);
    console.log(`Genres:    ${GENRES.join(", ")}`);
    return;
  }

  if (opts.status) {
    const result = await pollStatus(opts.status);
    const exp = result.export || {};
    const url = exp.dubbed_video_url || "";
    if (url) {
      console.log(`  URL: ${url}`);
      if (!opts.noDownload) {
        await downloadFile(url, opts.output || `dubbed_${opts.status}.mp4`);
      }
    }
    return;
  }

  if (!opts.video) { printHelp(); process.exit(1); }
  if (!opts.targetLang) {
    console.error("Error: --target-lang / -t is required");
    process.exit(1);
  }
  if (!fs.existsSync(opts.video)) {
    console.error(`Error: File not found: ${opts.video}`);
    process.exit(1);
  }

  const targetLangs = opts.targetLang.split(",").map((l) => l.trim());
  for (const lang of [...targetLangs, opts.srcLang]) {
    if (!LANGUAGES.includes(lang)) {
      console.error(`Error: '${lang}' is not valid.`);
      console.error(`Valid: ${LANGUAGES.join(", ")}`);
      process.exit(1);
    }
  }

  const jobName = opts.jobName || path.parse(opts.video).name;

  // 1. Create
  const jobData = await createJob(opts.srcLang, targetLangs, jobName, opts.speakers, opts.genre);

  // 2. Upload
  await upload(jobData.upload_url, opts.video);

  // 3. Start
  await startJob(jobData.job_id);

  // 4. Wait
  const result = await pollStatus(jobData.job_id);

  // 5. Download
  const exp = result.export || {};
  const dubbedUrl = exp.dubbed_video_url || "";

  if (dubbedUrl) {
    if (opts.noDownload) {
      console.log(`\nURL: ${dubbedUrl}`);
    } else {
      const outputPath = opts.output ||
        `${path.parse(opts.video).name}_${targetLangs.join("_")}${path.extname(opts.video)}`;
      await downloadFile(dubbedUrl, outputPath);
    }
    console.log(`\nDone! Job ID: ${jobData.job_id}`);
    if (exp.duration) console.log(`Duration: ${exp.duration.toFixed(1)}s`);
  } else {
    console.log(`\nNo output URL. Check: node dub.js --status ${jobData.job_id}`);
  }
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});

File: dub.py
#!/usr/bin/env python3
"""
Sarvam AI Video Dubbing CLI Tool
No authentication required!

Usage:
    python dub.py video.mp4 -t Hindi
    python dub.py video.mp4 --src-lang English --target-lang Hindi,Tamil
    python dub.py video.mp4 -t Telugu --speakers 2 --genre podcast
"""

import argparse
import os
import sys
import time
from pathlib import Path

import requests

BASE_URL = "https://dashboard.sarvam.ai"
API_DUBBING = f"{BASE_URL}/api/dubbing"

SUPPORTED_LANGUAGES = [
    "Assamese", "Bengali", "English", "Gujarati", "Hindi",
    "Kannada", "Malayalam", "Marathi", "Odia", "Punjabi",
    "Tamil", "Telugu",
]

SUPPORTED_GENRES = [
    "podcast", "monologue", "advertisement",
    "ott_movie_sequence", "edtech", "academic_lecture_formal",
]

POLL_INTERVAL = 5


def create_job(src_lang, target_langs, job_name, num_speakers, genre):
    """Create a dubbing job and get upload URL."""
    payload = {
        "src_lang": src_lang,
        "target_langs": target_langs,
        "job_name": job_name,
        "num_speakers": num_speakers,
        "genre": genre,
        "editor_flow": False,
    }

    print(f"\nCreating dubbing job...")
    print(f"  Source:   {src_lang}")
    print(f"  Target:   {', '.join(target_langs)}")
    print(f"  Speakers: {num_speakers}")
    print(f"  Genre:    {genre}")

    resp = requests.post(f"{API_DUBBING}/jobs", json=payload)
    if resp.status_code != 200:
        print(f"Error creating job: {resp.status_code}")
        print(resp.text)
        sys.exit(1)

    data = resp.json()["data"]
    print(f"  Job ID:   {data['job_id']}")
    return data


def upload_video(upload_url, video_path):
    """Upload video to Azure Blob Storage."""
    file_size = os.path.getsize(video_path)
    print(f"\nUploading: {video_path} ({file_size / 1024 / 1024:.1f} MB)")

    with open(video_path, "rb") as f:
        resp = requests.put(
            upload_url,
            data=f,
            headers={"x-ms-blob-type": "BlockBlob"},
        )

    if resp.status_code in (200, 201):
        print("  Upload complete!")
    else:
        print(f"Error uploading: {resp.status_code}")
        print(resp.text[:500])
        sys.exit(1)


def start_job(job_id):
    """Start processing the uploaded video."""
    print(f"\nStarting job...")
    resp = requests.post(f"{API_DUBBING}/jobs/{job_id}/start")
    if resp.status_code != 200:
        print(f"Error starting job: {resp.status_code}")
        print(resp.text)
        sys.exit(1)
    print("  Processing started!")


def poll_status(job_id):
    """Poll job status until completion."""
    print(f"\nWaiting for dubbing to complete...")
    last_step = ""

    while True:
        resp = requests.get(f"{API_DUBBING}/jobs/{job_id}/live-status")
        if resp.status_code != 200:
            print(f"  Status check failed: {resp.status_code}")
            time.sleep(POLL_INTERVAL)
            continue

        data = resp.json().get("data", {})
        status = data.get("status", "unknown")
        progress = data.get("progress", 0)
        step_label = data.get("current_step_label", "")

        if step_label and step_label != last_step:
            print(f"  [{progress:3.0f}%] {step_label}")
            last_step = step_label

        if status == "completed":
            export_data = data.get("export", {})
            if export_data.get("status") == "completed":
                print(f"\n  Dubbing completed!")
                return data
            elif export_data.get("status") == "failed":
                print(f"\n  Export failed: {data.get('error_message', 'Unknown')}")
                sys.exit(1)

        if status == "failed":
            print(f"\n  Failed: {data.get('error_message', 'Unknown')}")
            sys.exit(1)

        if status == "partial_failure":
            print(f"\n  Partial failure: {data.get('error_message', 'Unknown')}")
            sys.exit(1)

        time.sleep(POLL_INTERVAL)


def download_video(dubbed_url, output_path):
    """Download the dubbed video."""
    print(f"\nDownloading dubbed video...")

    if "blob.core.windows.net" in dubbed_url:
        url = f"{BASE_URL}/api/media-proxy?url={requests.utils.quote(dubbed_url, safe='')}"
    else:
        url = dubbed_url

    resp = requests.get(url, stream=True)
    if resp.status_code != 200:
        resp = requests.get(dubbed_url, stream=True)
        if resp.status_code != 200:
            print(f"Download failed. URL: {dubbed_url}")
            return

    total = int(resp.headers.get("content-length", 0))
    downloaded = 0

    with open(output_path, "wb") as f:
        for chunk in resp.iter_content(chunk_size=8192):
            f.write(chunk)
            downloaded += len(chunk)
            if total:
                print(f"\r  {downloaded / total * 100:.1f}%", end="")

    print(f"\n  Saved: {output_path}")


def main():
    parser = argparse.ArgumentParser(
        description="Sarvam AI Video Dubbing - No auth required!",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=f"""
Languages: {', '.join(SUPPORTED_LANGUAGES)}
Genres:    {', '.join(SUPPORTED_GENRES)}

Examples:
  python dub.py video.mp4 -t Hindi
  python dub.py video.mp4 -t Hindi,Tamil,Telugu --genre podcast
  python dub.py video.mp4 -t Bengali --speakers 2
  python dub.py --status JOB_ID
        """,
    )
    parser.add_argument("video", nargs="?", help="Path to video file")
    parser.add_argument("--src-lang", "-s", default="English", help="Source language (default: English)")
    parser.add_argument("--target-lang", "-t", help="Target language(s), comma-separated")
    parser.add_argument("--speakers", type=int, default=1, help="Number of speakers (default: 1)")
    parser.add_argument("--genre", "-g", default="podcast", choices=SUPPORTED_GENRES, help="Video genre (default: podcast)")
    parser.add_argument("--job-name", help="Custom job name")
    parser.add_argument("--output", "-o", help="Output file path")
    parser.add_argument("--no-download", action="store_true", help="Just print the URL")
    parser.add_argument("--status", help="Check status of existing job ID")
    parser.add_argument("--list", "-l", action="store_true", help="List languages & genres")

    args = parser.parse_args()

    if args.list:
        print("\nLanguages:", ", ".join(SUPPORTED_LANGUAGES))
        print("Genres:   ", ", ".join(SUPPORTED_GENRES))
        return

    if args.status:
        result = poll_status(args.status)
        export_data = result.get("export", {})
        url = export_data.get("dubbed_video_url", "")
        if url:
            print(f"  URL: {url}")
            if not args.no_download:
                download_video(url, args.output or f"dubbed_{args.status}.mp4")
        return

    if not args.video:
        parser.print_help()
        sys.exit(1)

    if not args.target_lang:
        print("Error: --target-lang / -t is required")
        print(f"Options: {', '.join(SUPPORTED_LANGUAGES)}")
        sys.exit(1)

    if not os.path.exists(args.video):
        print(f"Error: File not found: {args.video}")
        sys.exit(1)

    # Parse & validate languages
    target_langs = [l.strip() for l in args.target_lang.split(",")]
    for lang in target_langs + [args.src_lang]:
        if lang not in SUPPORTED_LANGUAGES:
            print(f"Error: '{lang}' is not valid.")
            print(f"Valid: {', '.join(SUPPORTED_LANGUAGES)}")
            sys.exit(1)

    job_name = args.job_name or Path(args.video).stem

    # 1. Create job
    job_data = create_job(args.src_lang, target_langs, job_name, args.speakers, args.genre)
    job_id = job_data["job_id"]
    upload_url = job_data["upload_url"]

    # 2. Upload video
    upload_video(upload_url, args.video)

    # 3. Start processing
    start_job(job_id)

    # 4. Wait for completion
    result = poll_status(job_id)

    # 5. Download result
    export_data = result.get("export", {})
    dubbed_url = export_data.get("dubbed_video_url", "")

    if dubbed_url:
        if args.no_download:
            print(f"\nURL: {dubbed_url}")
        else:
            if args.output:
                output_path = args.output
            else:
                stem = Path(args.video).stem
                ext = Path(args.video).suffix
                output_path = f"{stem}_{'_'.join(target_langs)}{ext}"
            download_video(dubbed_url, output_path)

        duration = export_data.get("duration", 0)
        print(f"\nDone! Job ID: {job_id}")
        if duration:
            print(f"Duration: {duration:.1f}s")
    else:
        print(f"\nNo output URL. Check: python dub.py --status {job_id}")


if __name__ == "__main__":
    main()
